File: tui/src/scala/tui/widgets/BlockWidget.scala
package tui
package widgets

import tui.Style
import tui.internal.ranges._
import tui.internal.saturating._

/** Base widget to be used with all upper level ones. It may be used to display a box border around the widget and/or add a title.
  *
  * @param title
  *   Optional title place on the upper left of the block
  * @param title_alignment
  *   Title alignment. The default is top left of the block, but one can choose to place title in the top middle, or top right of the block
  * @param borders
  *   Visible borders
  * @param border_style
  *   Border style
  * @param border_type
  *   Type of the border. The default is plain lines but one can choose to have rounded corners or doubled lines instead.
  * @param style
  *   Widget style
  */
case class BlockWidget(
    title: Option[Spans] = None,
    title_alignment: Alignment = Alignment.Left,
    borders: Borders = Borders.NONE,
    border_style: Style = Style.DEFAULT,
    border_type: BlockWidget.BorderType = BlockWidget.BorderType.Plain,
    style: Style = Style.DEFAULT
) extends Widget {

  /** Compute the inner area of a block based on its border visibility rules.
    */
  def inner(area: Rect): Rect = {
    var inner = area
    if (borders.intersects(Borders.LEFT)) {
      inner = inner.copy(
        x = inner.x.saturating_add(1).min(inner.right),
        width = inner.width.saturating_sub_unsigned(1)
      )
    }
    if (borders.intersects(Borders.TOP) || title.isDefined) {
      inner = inner.copy(
        y = inner.y.saturating_add(1).min(inner.bottom),
        height = inner.height.saturating_sub_unsigned(1)
      )
    }
    if (borders.intersects(Borders.RIGHT)) {
      inner = inner.copy(width = inner.width.saturating_sub_unsigned(1))
    }
    if (borders.intersects(Borders.BOTTOM)) {
      inner = inner.copy(height = inner.height.saturating_sub_unsigned(1))
    }
    inner
  }

  def render(area: Rect, buf: Buffer): Unit = {
    if (area.area == 0) {
      return
    }
    buf.set_style(area, style)
    val symbols = BlockWidget.BorderType.line_symbols(border_type)

    // Sides
    if (borders.intersects(Borders.LEFT)) {
      range(area.top, area.bottom) { y =>
        buf
          .get(area.left, y)
          .set_symbol(symbols.vertical)
          .set_style(border_style)
        ()
      }
    }

    if (borders.intersects(Borders.TOP)) {
      range(area.left, area.right) { x =>
        buf
          .get(x, area.top)
          .set_symbol(symbols.horizontal)
          .set_style(border_style)
        ()
      }
    }
    if (borders.intersects(Borders.RIGHT)) {
      val x = area.right - 1
      range(area.top, area.bottom) { y =>
        buf
          .get(x, y)
          .set_symbol(symbols.vertical)
          .set_style(border_style)
        ()
      }
    }
    if (borders.intersects(Borders.BOTTOM)) {
      val y = area.bottom - 1
      range(area.left, area.right) { x =>
        buf
          .get(x, y)
          .set_symbol(symbols.horizontal)
          .set_style(border_style)
        ()
      }
    }

    // Corners
    if (borders.contains(Borders.RIGHT | Borders.BOTTOM)) {
      buf
        .get(area.right - 1, area.bottom - 1)
        .set_symbol(symbols.bottom_right)
        .set_style(border_style)
    }
    if (borders.contains(Borders.RIGHT | Borders.TOP)) {
      buf
        .get(area.right - 1, area.top)
        .set_symbol(symbols.top_right)
        .set_style(border_style)
    }
    if (borders.contains(Borders.LEFT | Borders.BOTTOM)) {
      buf
        .get(area.left, area.bottom - 1)
        .set_symbol(symbols.bottom_left)
        .set_style(border_style)
    }
    if (borders.contains(Borders.LEFT | Borders.TOP)) {
      buf
        .get(area.left, area.top)
        .set_symbol(symbols.top_left)
        .set_style(border_style)
    }

    // Title
    title.foreach { title =>
      val left_border_dx = if (borders.intersects(Borders.LEFT)) 1 else 0

      val right_border_dx = if (borders.intersects(Borders.RIGHT)) { 1 }
      else { 0 }

      val title_area_width = area.width
        .saturating_sub_unsigned(left_border_dx)
        .saturating_sub_unsigned(right_border_dx)

      val title_dx = title_alignment match {
        case Alignment.Left   => left_border_dx
        case Alignment.Center => area.width.saturating_sub_unsigned(title.width) / 2
        case Alignment.Right =>
          area.width
            .saturating_sub_unsigned(title.width)
            .saturating_sub_unsigned(right_border_dx)
      }

      val title_x = area.left + title_dx
      val title_y = area.top

      buf.set_spans(title_x, title_y, title, title_area_width);
    }
  }
}

object BlockWidget {
  sealed trait BorderType

  object BorderType {
    case object Plain extends BorderType

    case object Rounded extends BorderType

    case object Double extends BorderType

    case object Thick extends BorderType

    def line_symbols(border_type: BorderType): symbols.line.Set =
      border_type match {
        case BorderType.Plain   => symbols.line.NORMAL
        case BorderType.Rounded => symbols.line.ROUNDED
        case BorderType.Double  => symbols.line.DOUBLE
        case BorderType.Thick   => symbols.line.THICK
      }
  }

}

File: tui/src/java/tui/widgets/BlockWidget.java
package tui.widgets;

import java.util.Optional;
import tui.Alignment;
import tui.Borders;
import tui.Buffer;
import tui.Rect;
import tui.Spans;
import tui.Style;
import tui.Symbols;
import tui.Widget;
import tui.internal.Ranges;
import tui.internal.Saturating;

/// Base widget to be used with all upper level ones.
public final class BlockWidget implements Widget {
  public final Optional<Spans> title;
  public final Alignment titleAlignment;
  public final Borders borders;
  public final Style borderStyle;
  public final BorderType borderType;
  public final Style style;

  public BlockWidget(
      Optional<Spans> title,
      Alignment titleAlignment,
      Borders borders,
      Style borderStyle,
      BorderType borderType,
      Style style) {
    this.title = title;
    this.titleAlignment = titleAlignment;
    this.borders = borders;
    this.borderStyle = borderStyle;
    this.borderType = borderType;
    this.style = style;
  }

  public static BlockWidget empty() {
    return new BlockWidget(
        Optional.empty(),
        Alignment.Left,
        Borders.NONE,
        Style.empty(),
        BorderType.Plain,
        Style.empty());
  }

  /// Compute the inner area of a block based on its border visibility rules.
  public Rect inner(Rect area) {
    Rect inner = area;
    if (borders.intersects(Borders.LEFT)) {
      inner =
          new Rect(
              Math.min(Saturating.saturatingAdd(inner.x(), 1), inner.right()),
              inner.y(),
              Saturating.saturatingSubUnsigned(inner.width(), 1),
              inner.height());
    }
    if (borders.intersects(Borders.TOP) || title.isPresent()) {
      inner =
          new Rect(
              inner.x(),
              Math.min(Saturating.saturatingAdd(inner.y(), 1), inner.bottom()),
              inner.width(),
              Saturating.saturatingSubUnsigned(inner.height(), 1));
    }
    if (borders.intersects(Borders.RIGHT)) {
      inner =
          new Rect(
              inner.x(),
              inner.y(),
              Saturating.saturatingSubUnsigned(inner.width(), 1),
              inner.height());
    }
    if (borders.intersects(Borders.BOTTOM)) {
      inner =
          new Rect(
              inner.x(),
              inner.y(),
              inner.width(),
              Saturating.saturatingSubUnsigned(inner.height(), 1));
    }
    return inner;
  }

  @Override
  public void render(Rect area, Buffer buf) {
    if (area.area() == 0) {
      return;
    }
    buf.setStyle(area, style);
    Symbols.line.Set symbols = BorderType.lineSymbols(borderType);

    // Sides
    if (borders.intersects(Borders.LEFT)) {
      Ranges.range(
          area.top(),
          area.bottom(),
          y -> buf.get(area.left(), y).setSymbol(symbols.vertical()).setStyle(borderStyle));
    }

    if (borders.intersects(Borders.TOP)) {
      Ranges.range(
          area.left(),
          area.right(),
          x -> buf.get(x, area.top()).setSymbol(symbols.horizontal()).setStyle(borderStyle));
    }
    if (borders.intersects(Borders.RIGHT)) {
      int rx = area.right() - 1;
      Ranges.range(
          area.top(),
          area.bottom(),
          y -> buf.get(rx, y).setSymbol(symbols.vertical()).setStyle(borderStyle));
    }
    if (borders.intersects(Borders.BOTTOM)) {
      int by = area.bottom() - 1;
      Ranges.range(
          area.left(),
          area.right(),
          x -> buf.get(x, by).setSymbol(symbols.horizontal()).setStyle(borderStyle));
    }

    // Corners
    if (borders.contains(Borders.RIGHT.or(Borders.BOTTOM))) {
      buf.get(area.right() - 1, area.bottom() - 1)
          .setSymbol(symbols.bottomRight())
          .setStyle(borderStyle);
    }
    if (borders.contains(Borders.RIGHT.or(Borders.TOP))) {
      buf.get(area.right() - 1, area.top()).setSymbol(symbols.topRight()).setStyle(borderStyle);
    }
    if (borders.contains(Borders.LEFT.or(Borders.BOTTOM))) {
      buf.get(area.left(), area.bottom() - 1)
          .setSymbol(symbols.bottomLeft())
          .setStyle(borderStyle);
    }
    if (borders.contains(Borders.LEFT.or(Borders.TOP))) {
      buf.get(area.left(), area.top()).setSymbol(symbols.topLeft()).setStyle(borderStyle);
    }

    // Title
    if (title.isPresent()) {
      Spans titleVal = title.get();
      int leftBorderDx = borders.intersects(Borders.LEFT) ? 1 : 0;
      int rightBorderDx = borders.intersects(Borders.RIGHT) ? 1 : 0;

      int titleAreaWidth =
          Saturating.saturatingSubUnsigned(
              Saturating.saturatingSubUnsigned(area.width(), leftBorderDx), rightBorderDx);

      int titleDx = switch (titleAlignment) {
        case Left -> leftBorderDx;
        case Center -> Saturating.saturatingSubUnsigned(area.width(), titleVal.width()) / 2;
        case Right -> Saturating.saturatingSubUnsigned(
            Saturating.saturatingSubUnsigned(area.width(), titleVal.width()), rightBorderDx);
      };

      int titleX = area.left() + titleDx;
      int titleY = area.top();

      buf.setSpans(titleX, titleY, titleVal, titleAreaWidth);
    }
  }

  public BlockWidget withTitle(Spans t) {
    return new BlockWidget(Optional.of(t), titleAlignment, borders, borderStyle, borderType, style);
  }

  public BlockWidget withBorders(Borders b) {
    return new BlockWidget(title, titleAlignment, b, borderStyle, borderType, style);
  }

  public BlockWidget withBorderStyle(Style s) {
    return new BlockWidget(title, titleAlignment, borders, s, borderType, style);
  }

  public BlockWidget withBorderType(BorderType t) {
    return new BlockWidget(title, titleAlignment, borders, borderStyle, t, style);
  }

  public BlockWidget withStyle(Style s) {
    return new BlockWidget(title, titleAlignment, borders, borderStyle, borderType, s);
  }

  public BlockWidget withTitleAlignment(Alignment a) {
    return new BlockWidget(title, a, borders, borderStyle, borderType, style);
  }

  public enum BorderType {
    Plain,
    Rounded,
    Double,
    Thick;

    public static Symbols.line.Set lineSymbols(BorderType borderType) {
      return switch (borderType) {
        case Plain -> Symbols.line.NORMAL;
        case Rounded -> Symbols.line.ROUNDED;
        case Double -> Symbols.line.DOUBLE;
        case Thick -> Symbols.line.THICK;
      };
    }
  }
}
